src/modules/job-info/update.c

```diff
@@ -0,0 +1,490 @@
+#include "update.h"
+#include "util.h"
+
+struct update_watch_ctx {
+    struct info_ctx *ctx;
+    const flux_msg_t *msg;
+    struct flux_msg_cred cred;
+    flux_jobid_t id;
+    char *key;
+    int flags;
+    const char *update_name;
+    flux_future_t *lookup_f;
+    flux_future_t *eventlog_watch_f;
+    json_t *update_object;
+    int initial_update_count;
+    int watch_update_count;
+    bool cancel;
+};
+
+static void update_watch_ctx_destroy (void *data)
+{
+    if (data) {
+        struct update_watch_ctx *uw = data;
+        int save_errno = errno;
+        flux_msg_decref (uw->msg);
+        free (uw->key);
+        flux_future_destroy (uw->lookup_f);
+        flux_future_destroy (uw->eventlog_watch_f);
+        json_decref (uw->update_object);
+        free (uw);
+        errno = save_errno;
+    }
+}
+
+static struct update_watch_ctx *update_watch_ctx_create (struct info_ctx *ctx,
+                                                         const flux_msg_t *msg,
+                                                         flux_jobid_t id,
+                                                         const char *key,
+                                                         int flags)
+{
+    struct update_watch_ctx *uw = calloc (1, sizeof (*uw));
+
+    if (!uw)
+        return NULL;
+
+    uw->ctx = ctx;
+    uw->id = id;
+    if (!(uw->key = strdup (key))) {
+        errno = ENOMEM;
+        goto error;
+    }
+    if (streq (key, "R"))
+        uw->update_name = "resource-update";
+    else {
+        errno = EINVAL;
+        goto error;
+    }
+    uw->flags = flags;
+
+    uw->msg = flux_msg_incref (msg);
+
+    if (flux_msg_get_cred (msg, &uw->cred) < 0) {
+        flux_log_error (ctx->h, "%s: flux_msg_get_cred", __FUNCTION__);
+        goto error;
+    }
+
+    return uw;
+
+error:
+    update_watch_ctx_destroy (uw);
+    return NULL;
+}
+
+static void eventlog_watch_cancel (struct update_watch_ctx *uw)
+{
+    flux_future_t *f;
+    int matchtag;
+
+    /* in some cases, possible eventlog watch hasn't started yet */
+    if (!uw->eventlog_watch_f)
+        return;
+
+    matchtag = (int)flux_rpc_get_matchtag (uw->eventlog_watch_f);
+
+    if (!(f = flux_rpc_pack (uw->ctx->h,
+                             "job-info.eventlog-watch-cancel",
+                             FLUX_NODEID_ANY,
+                             FLUX_RPC_NORESPONSE,
+                             "{s:i}",
+                             "matchtag", matchtag))) {
+        flux_log_error (uw->ctx->h, "%s: flux_rpc_pack", __FUNCTION__);
+        return;
+    }
+    flux_future_destroy (f);
+}
+
+static void apply_updates_R (struct update_watch_ctx *uw,
+                             json_t *context)
+{
+    const char *key;
+    json_t *value;
+
+    json_object_foreach (context, key, value) {
+        /* RFC 21 resource-update event only allows update
+         * to:
+         * - execution.expiration
+         */
+        if (!streq (key, "execution.expiration"))
```

Comment:
Similar to the other PR, the actual key is just `expiration`

Reply:
interesting, I now see the `resource-update` context is formatted differently in RFC21 than `jobspec-update`.

Did we have a specific reason for that?  It seems possibly unwise long term for both contexts to be formatted differently?  On the other hand I can understand that `R` should have far more limited update possibilities.

Reply:
> Did we have a specific reason for that? It seems possibly unwise long term for both contexts to be formatted differently? 

I think it was done this way because jobspec and R are actually quite different - R is not a simple hierarchical document like jobspec. E.g., when grow/shrink or similar are added, it will be more useful to have the added or deleted R and not a wholesale replacement of `execution.R_lite`.